### PR TITLE
Ticket #687: Formatting for nameservers

### LINKED ISSUE #687: [Bug]: Nameserver display isn't correct for new domain interface
### Current Behavior

The new domain interface that we are building has changed the return type of the `nameservers()` method. The domain overview page displays the name servers in a strange way because of the change in that method.

Screenshot 2023-06-07 at 9 46 25 AM

### Expected Behavior

The list of nameservers should not include any parentheses or commas.

### Steps to Reproduce

1. Visit the approved domain overview page at /domain/<id>

### Environment (optional)

_No response_

### Additional Context (optional)

_No response_

### PR DESCRIPTION
**Pull request by therealslimhsiehdy**
## Ticket

Resolves #687 

## Changes
* Format DNS nameserver string to: NameserverName (IP addresses) or NameserverName when there are no IP addresses listed.

## Context for reviewers

We want to fix how we display nameserver and ip.

Previously, it looked like:
* Just nameserver, no IP: `('ns1.helloerin.com', [])`
* Has nameserver and IP: `('ns1.helloerin.com', [89.89.89.89])`

We want it to display like so:
* Just nameserver, no IP: `ns1.helloerin.com`
* Has nameserver and IP: `ns1.helloerin.com (89.89.89.89)`

Please note that cache is currently returning epp.Ip object hence why you'll see in the screenshot the IP object header.
Previous: `[IP(addr='127.00.0.1', ip='v4'), Ip(addr='::1', ip='v6')]`
Current: `(IP(addr='127.00.0.1', ip='v4'), Ip(addr='::1', ip='v6'))`

## Setup

Log into your own sandbox + reference `registry-access.md` for how to login + imports. 

Commands to get into sandbox shell:

```
cf ssh getgov-[YOURENVHERE]
/tmp/lifecycle/shell
./manage.py shell
```

Import EPP code:

```
from epplibwrapper import CLIENT as registry, commands
from epplib.models import common
```

Create a nameserver for the domain you have:

```
request = commands.CreateHost(name="ns17.note-side-threat.gov", addrs=[common.Ip(addr="127.9.9.1"), common.Ip(addr="0:0:0:0:0:0:0:1", ip="v6")])
response = registry.send(request, cleaned=True)
request = commands.UpdateDomain(name="note-side-threat.gov", add=[common.HostObjSet(["ns17.note-side-threat.gov"])])
response = registry.send(request, cleaned=True)
```

(You can print the response with `print(response)` to make sure you get a 200 and that the requests are successful)

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [x] Pulled this branch locally and tested it
- [x] Reviewed this code and left comments
- [x] Checked that all code is adequately covered by tests
- [x] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [x] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

## Screenshots

image

### CODE REVIEW

File: src/registrar/models/domain.py [hunk 1]
```diff
@@ -228,6 +228,44 @@ def password(self) -> str:
         """
         raise NotImplementedError()
 
+    @Cache
```
Comment (abroddrick) [inline src/registrar/models/domain.py:231]: why @Cache here that sets it as property?

File: src/registrar/templates/domain_detail.html [hunk 2]
```diff
@@ -28,8 +28,8 @@
     <br>
 
     {% url 'domain-dns-nameservers' pk=domain.id as url %}
-    {% if domain.nameservers|length > 0 %}
-      {% include "includes/summary_item.html" with title='DNS name servers' value=domain.nameservers list='true' edit_link=url %}
+    {% if domain.format_nameservers|length > 0 %}
```
Comment (abroddrick) [inline src/registrar/templates/domain_detail.html:31]: why are we changing it here and not the summary item element? there is a todo in that with this ticket # that says this is where the ticket needs to be fixed. @therealslimhsiehdy do you remember us adding that todo when we did nameservers, or was this intentional?

File: src/registrar/tests/common.py [hunk 3]
```diff
@@ -804,7 +805,20 @@ def dummyInfoContactResultData(
     infoDomainHasIP = fakedEppObject(
         "nameserverwithip.gov",
         cr_date=datetime.datetime(2023, 5, 25, 19, 45, 35),
-        contacts=[],
+        contacts=[
+            common.DomainContact(
```
Comment (abroddrick) [inline src/registrar/tests/common.py:809]: why did you add this?

File: src/registrar/models/domain.py [hunk 4]
```diff
@@ -228,6 +228,44 @@ def password(self) -> str:
...
+    def format_nameservers(self):
+        """
```
Comment (abroddrick) [inline src/registrar/models/domain.py:233]: So this does violate the MVT set up, your model ideally shouldn't dictate any logic with formatting. The model should just handle CRUD operations on the data itself, while View is for data processing pre/post retrieval as well as http/rendering, and then the template should be your format for your page. Ideally this should all go in the template, however as it does involve data manipulation for the sole purpose of display this sounds like you can just cart this over to the view instead.

File: src/registrar/models/domain.py [hunk 5]
```diff
@@ -228,6 +228,44 @@ def password(self) -> str:
...
+            return []
+
+        hostList = []
+        for host in hosts:
+            # can remove str conversion when EPP.IP address display
+            # changes from IP object -> IP address
+            host_addrs_stringified = [str(addr) for addr in host["addrs"]]
+            host_addrs_str = f'({", ".join(obj for obj in host_addrs_stringified)})'
+            host_info_str = f'{host["name"]} \
+            {host_addrs_str if len(host["addrs"]) > 0 else ""}'
```
Comment (abroddrick) [inline src/registrar/models/domain.py:265]: logic checks out just think it needs to move to the view

File: src/registrar/templates/domain_detail.html (hunk 2 above)
Comment (therealslimhsiehdy) [reply src/registrar/templates/domain_detail.html:31]: GREAT catch, I saw it in the nameserver function, but completely forgot about it in the `includes/summary_item.html` file. Will remedy this!

File: src/registrar/tests/common.py (hunk 3 above)
Comment (therealslimhsiehdy) [reply src/registrar/tests/common.py:809]: For the tests it kept erroring out saying we needed Contacts, with this error `raise ContactError("No contact was found in cache or the registry")`

Rachid, Erin and I thought this was odd too...

File: src/registrar/models/domain.py (hunk 4 above)
Comment (therealslimhsiehdy) [reply src/registrar/models/domain.py:233]: Yes 100%!! This was totally my bad I smooth brained and put it the file and didn't even think about logic vs view side. Will remedy this! Thank you for catching this, I totally borked on this 😵‍💫

File: src/registrar/tests/common.py [hunk 6]
```diff
@@ -813,6 +827,29 @@ def dummyInfoContactResultData(
         addrs=["1.2.3.4", "2.3.4.5"],
```
Comment (dave-kennedy-ecs) [inline src/registrar/tests/common.py:827]: Reminder that the format of addrs will need to be updated once issue 1016 is merged.

**Comment by gabydisarli** [comment]
I'm still sort of confused as to how these will look once the caching issue is resolved. Can you expand on the formatting we're after?

File: src/registrar/tests/common.py (hunk 3 above)
Comment (abroddrick) [reply src/registrar/tests/common.py:798]: That is annoying....

**Comment by abroddrick** [comment]
> I'm still sort of confused as to how these will look once the caching issue is resolved. Can you expand on the formatting we're after?

@gabydisarli it will be `ns1.example.com (12.123.1234, 89.89.89.89 )` where the values in the parenthesis are the ips. It does look like the text will wrap to the next line if there is a lot of ips (not sure how common that is)

**Review by abroddrick — APPROVED** [review]
Thanks for making changes. Up to Gaby on if spacing looks right though. It also looks like we may want to handle formatting for if it has a lot of ips, but doesn't seem critical

**Comment by gabydisarli** [comment]
@therealslimhsiehdy That sounds good and what I was expecting. I was thrown off by this:

> We want it to display like so:
> 
> Just nameserver, no IP: ns1.helloerin.com
> Has nameserver and IP: ns1.helloerin.com [89.89.89.89])

I was not understanding the mix of bracket and parenthesis use here but if it is like you said above `ns1.example.com (12.123.1234, 89.89.89.89)` Then that sounds good. I think because the field is short a user wouldn't be adding more than maybe 3 IPs and would instead resort to using another field set.

**Review by gabydisarli — APPROVED** [review]
I wasn't able to see this on your sandbox but from the comments I think the formatting looks good. Thanks Rebecca!

**Comment by therealslimhsiehdy** [comment]
> @therealslimhsiehdy That sounds good and what I was expecting. I was thrown off by this:
> 
> > We want it to display like so:
> > Just nameserver, no IP: ns1.helloerin.com
> > Has nameserver and IP: ns1.helloerin.com [89.89.89.89])
> 
> I was not understanding the mix of bracket and parenthesis use here but if it is like you said above `ns1.example.com (12.123.1234, 89.89.89.89)` Then that sounds good. I think because the field is short a user wouldn't be adding more than maybe 3 IPs and would instead resort to using another field set.

@gabydisarli Sorry that was my bad in the description, just edited it -- it should look like `ns1.helloerin.com (89.89.89.89, 12.1234.12.1234)` with the new code! Unfortunately there's a few extra steps involved in order to get an ip address to display on sandbox (you'll have to manually add it in), but I did attach a screenshot in case :-) Thank you for all your help Gaby ❤️
